pods/pods.go

```diff
@@ -33,127 +89,125 @@ const (
 }
 
-var testData []byte
+func (c *ContainerStatus) Equal(oc *ContainerStatus) bool {
+	return !(c.Name != oc.Name ||
+		c.Image != oc.Image ||
+		c.Ready != oc.Ready ||
+		c.Reason != oc.Reason ||
+		c.Message != oc.Message)
+}
 
 // GetPodStatus returns the status of the pods found in the supplied namespace.
-func GetPodStatus(namespace string) ([]PodStatus, error) {
-	var data []byte
-	if testData == nil {
-		var stdout bytes.Buffer
-		args := []string{"get", "pods", "-o", "json"}
-		if namespace != "" {
-			args = append(args, "-n", namespace)
-		} else {
-			args = append(args, "-A")
-		}
-		cmd := exec.Command("kubectl", args...)
-		cmd.Stdout = &stdout
-		cmd.Stderr = os.Stderr
-		if err := cmd.Run(); err != nil {
-			return nil, err
-		}
-		data = stdout.Bytes()
-	} else {
-		data = testData
-	}
-
-	var pods corev1.PodList
-	var statuses []PodStatus
-	if err := json.Unmarshal(data, &pods); err != nil {
+func GetPodStatus(ctx context.Context, client kubernetes.Interface, namespace string) ([]*PodStatus, error) {
+	pods, err := client.CoreV1().Pods(namespace).List(ctx, metav1.ListOptions{})
+	if err != nil {
 		return nil, err
 	}
+	var statuses []*PodStatus
 	for _, pod := range pods.Items {
-		statuses = append(statuses, podStatus(pod))
+		pod := pod // to keep the linter happy
+		statuses = append(statuses, PodToStatus(&pod))
 	}
 	return statuses, nil
 }
 
-var testReader io.Reader
-
 // WatchPodStatus returns a channel on which the status of the pods in the
-// supplied namespace are written.  If quit is closed then WatchPodSatus's go
-// routine will exit.  Passing nil as quit will cause the go routine to only
-// exit if there is an error watching the pods.
-func WatchPodStatus(namespace string, quit chan struct{}) (chan PodStatus, error) {
-	var r io.Reader
-	if testReader == nil {
-		args := []string{"get", "pods", "-o", "json", "--watch"}
-		if namespace != "" {
-			args = append(args, "-n", namespace)
-		} else {
-			args = append(args, "-A")
-		}
-		fmt.Printf("args: %q\n", args)
-		cmd := exec.Command("kubectl", args...)
-		stdout, err := cmd.StdoutPipe()
-		cmd.Stderr = os.Stderr
-		if err != nil {
-			return nil, err
-		}
-		if err := cmd.Start(); err != nil {
-			return nil, err
-		}
-		go cmd.Wait()
-		r = stdout
-	} else {
-		r = testReader
+// supplied namespace are written.
+func WatchPodStatus(ctx context.Context, client kubernetes.Interface, namespace string) (chan *PodStatus, func(), error) {
+	// Make sure we can even get the pods
+	w, err := client.CoreV1().Pods(namespace).Watch(ctx, metav1.ListOptions{})
+	if err != nil {
+		return nil, nil, err
 	}
-	ch := make(chan PodStatus, 2)
+	kch := w.ResultChan()
+	ch := make(chan *PodStatus, 2)
+
 	go func() {
 		defer close(ch)
-		d := json.NewDecoder(r)
-		for {
-			var pod corev1.Pod
-			if err := d.Decode(&pod); err != nil {
-				if err != io.EOF {
-					select {
-					case <-quit:
-					case ch <- PodStatus{Error: err}:
-					}
+		// seen keeps track of
```

Comment:
finish comment, not sure what seen is tracking, would we not want the status to update even if we've seen the pod UID before, or maybe thats the event UID?

Reply:
bump

Reply:
Overcome by events.

Reply:
Oh, the change wasn't pushed yet.